src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,115 @@
+        // this flag is needed to know if ancestor replaces very old descentdant in the tree
+        let descendantReplaced = false // eslint-disable-line fp/no-let
+        // If A.E which already available in tree is updated to A.EM the updating just its descendants like A.E.O.M to A.EM.O.M.
+        leafNodes.forEach(descendant => {
+          if (timeDifference(timestamp(), descendant.lastUpdated) > editTimeMax) {
+            _.unset(tree, findClosestAncestorWithMultipleChildren(tree, ['_ROOT_', ...pathToContext(descendant.path)]).path)
+            descendantReplaced = true
+          }
+          else {
+            const updatedDescendantPath = newPath.concat(descendant.path.slice(newPath.length))
+            _.unset(tree, ['_ROOT_', ...pathToContext(descendant.path)])
+            _.set(tree, ['_ROOT_', ...pathToContext(updatedDescendantPath)], { leaf: true, lastUpdated: timestamp(), path: updatedDescendantPath })
+          }
+        })
+        if (descendantReplaced) onNodeChange(tree, oldPath, newPath) // called once again to remove merge inconsitencty that might occur while replacing descendants by ancestor
+        else _.unset(tree, oldContext)
+      }
+      else {
+        let isMerged = false // eslint-disable-line fp/no-let
+        leafNodes.forEach(descendant => {
+          const descendantContext = ['_ROOT_', ...pathToContext(descendant.path)]
+          if (descendantContext[1] === '__EM__') return // preventing thoughts at level 1 from merged to this (temporary fix)
+          const longContext = descendantContext.length > newContext.length ? descendantContext : newContext
+          const shortContext = newContext.length < descendantContext.length ? newContext : descendantContext
+          // siblings A.B.C and A.B.D ---> A.B.D
+          if (descendantContext.length === newContext.length && newContext.length - commonPath.length === 1) {
+            isMerged = true
+            _.unset(tree, descendantContext)
+            _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+          }
+          /*
+            restricting merge of direct cousins like A.B.C.F and A.B.D.E
+            merging distant relation A.B.C and A.B.D.E.F.G.H ---> A.B.D.E.F.G.H
+          */
+          else if (!(shortContext.length - commonPath.length > 1 || (newContext.length === descendantContext.length && newContext.length - commonPath.length === 2))) {
+            isMerged = true
+            _.unset(tree, shortContext.slice(0, commonPath.length + 1))
+            _.set(tree, longContext, { leaf: true, lastUpdated: timestamp(), path: newContext.length > descendantContext.length ? newPath : descendant.path })
+          }
+        })
+        // adding new thought to the tree
+        if (!isMerged) _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+      }
+    }
+  }
+}
+
+export const onNodeDelete = (tree, oldPath) => {
```

Comment:
Let's call these methods `treeDelete` and `treeChange`. `onSomething` indicates an event handler, which is not what this is.